src/boot/efifirmware.c

```diff
@@ -0,0 +1,86 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "efifirmware.h"
+#include "util.h"
+#include <endian.h>
+
+static bool efifw_validate_header(
+                const void *blob,
+                size_t blob_len,
+                const char **ret_fwid_off,
+                const char **ret_payload_off) {
+
+        if ((uintptr_t) blob % alignof(EfiFwHeader) != 0)
+                return false;
+
+        size_t base_sz = sizeof_field(EfiFwHeader, magic) +
+                sizeof_field(EfiFwHeader, header_len) +
+                sizeof_field(EfiFwHeader, fwid_len) +
+                sizeof_field(EfiFwHeader, payload_len);
+        /* at least the base size of the header must be in memory */
+        if (blob_len < base_sz)
+                return false;
+
+        const EfiFwHeader *fw_header = ASSERT_PTR(blob);
+
+        if (fw_header->magic != FWHEADERMAGIC)
+                return false;
+
+        uint32_t header_len  = fw_header->header_len;
+
+        /* header_len must not be malformed */
+        if (header_len < base_sz)
+                return false;
+
+        uint32_t fwid_len    = fw_header->fwid_len;
+        uint32_t payload_len = fw_header->payload_len;
+        size_t total_computed_size;
+
+        /* check for unusually large values of payload_len, header_len or fwid_len */
+        if (!ADD_SAFE(&total_computed_size, header_len, fwid_len) ||
+            !ADD_SAFE(&total_computed_size, total_computed_size, payload_len) )
+                return false;
+
+        /* see if entire size of the base header is present in memory */
+        if (blob_len < total_computed_size)
+                return false;
+
+        char *fwid_off    = (char*)blob + header_len;
+        char *payload_off = fwid_off + fwid_len;
+
+        /* check that fwid_off contains NULL character within fwid_len */
+        if (!memchr(fwid_off, 0, fwid_len))
```

Comment:
the check is actually a bit too strict and also too lose... we want to check that there's exactly one NUL byte, and that it is the last byte in the buffer.

i.e. 
```c
if (memchr(fwid, 0, fwid_len) != fwd + fwid_len - 1)
        return false;
```